### PR TITLE
🐛 Source HubSpot: fix infinite loop when iterating through search results

### PR DESCRIPTION
**Pull request by ehearty**
🐛 Source HubSpot: fix infinite loop when iterating through search results

## What
Fixes airbytehq/airbyte/#43317

## How
I updated our custom HS connector to filter on both primary key AND timestamp, then to sort by object's primary key. The timestamps new return out of order, but once we finish iterating through all the results we should have all the data (similar to full refresh processing).

## Review guide
1. `airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py`
     - updates to the "process_search" function:
        - added a "last_id" parameter 
        - updated filter logic to include:
           - self.primary_key >= last_id
           - self.last_modified_field >= self._state.timestamp()
           - self.last_modified_field <= self._init_sync.timestamp() 
              *note: I added this logic because I don't like "open ended queries" where the result set might change depending on whether new records were modified after the start of the pull.*
        - updated sort logic to use the primary key instead of the last_modified_field to ensure that we're always grabbing a unique result set
     - added a "get_max" function that will first attempt a numeric comparison, but fall back to a string comparison if int conversion isn't possible - this is mostly to account for unit tests where the "id" field is a string instead of an integer, and any future object searches where we might want to use a non-numeric primary key
     - updated read_records to:
        - grab the maximum primary key returned in the result set, and pass that as the last_id value to the process_search function every time we go over 10k results
        - no longer update the _state with the last_modified_timestamp now that we're sorting by primary key instead
        - once the search completes, update the state with the init_sync_timestamp to ensure that the next run starts from exactly where we left off
## User Impact
* What is the end result perceived by the user? None.
* If there are negative side effects, please list them. 
  - Now that we're no longer updating the state with the last modified date after each iteration, we'll lose any partial progress if an error occurs during the sync. Given how rarely this scenario occurs, and how quickly and easily we can re-pull this data, impact should be minimal.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by ehearty — COMMENTED** [review]
closes airbytehq/airbyte/#43317

**Comment by ehearty** [comment]
@netoferraz  @aldogonzalez8 - I'm not sure if you guys are the right people to reach out to on this, but I'm seeing the `Check if a review is required from Connector teams on fork` check fail and I can't seem to find any documentation about what it does. Could someone please point me in the right direction? 

Also, it's been a loooong time since I linked an issue to a PR in github - please let me know if I did that correctly or if I need to update something else on my end.

Best,
Elena Hearty

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (aldogonzalez8) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1217]: Left over from debugging?

Comment (ehearty) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1217]: good catch! updating now...

Comment (ehearty) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py]: @aldogonzalez8 just wanted to let you know that I updated the PR to remove that print statement. Let me know if you spot anything else.

Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py]: Hey, sorry, need to fix HS CAT as one is broken because of missing data in test account before we can approve and merge, will let you know when is done.

**Comment by aldogonzalez8** [comment]
I will put this here as a comment just if in the future, somebody was to understand the change and has some inquiries:

Before:

1. Request data to HS on the search endpoint 
   - filter on the latest state (timestamp).
   - ordered. by timestamp
2. We use the token returned in response to paginate.
3. If the token is none, pagination is over
4. If pagination >10000 records, we start a new request as we can't get further because HS doesn't want that.
5. New request will use the latest/greatest timestamp
6. if all data has identical timestamps because some batch update occurred, a new request would try to use the same initial timestamp, and we get the infinite loop.

After:

1. Request data on HS in the search endpoint:
   - filter on the latest state (timestamp)
   - Filter on latest object id/primary_key with initial None
   - Filter on timestamp < current_sync to not have "open-ended queries"
2. We use the token returned in response to paginate.
3. If the token is none, pagination is over
4. If pagination > 1000 records, we start a new request because we can't go further, as HS doesn't want that.
5. New requests will filter like:
   - **Filter still on the latest state before sync started**
   - Use the latest object ID/primary_key obtained from the previous pagination process (technically, it's the same loop, so pagination is not over, but we restart because of the 10,000 limit and that stuff.).
   - Again, Filter on timestamp < current_sync to not have "open-ended queries"
6. if all data has identical timestamps because some batch update occurred, this will not affect the pagination process as we also filtered and now instead sorted by object ID rather than timestamp.
9. We are happy because it is not an infinite loop.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (aldogonzalez8) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1142]: @ehearty, Would you mind putting here some documentation about why this may be needed? It doesn't need to be as large as my comment in the PR, but some TL;DR on why we have a second filter may be complementary to the one in the read records section that talks about the 10,000 limitations.

My only concern is that we don't have a specific test for this scenario. Would you mind taking a look at the chance to have one? Please let me know what you think.

File: docs/integrations/sources/hubspot.md
Comment (aldogonzalez8) [inline docs/integrations/sources/hubspot.md:335]: We need to update the date here.

**Comment by aldogonzalez8** [comment]
LGTM left some comments, but I'm happy the unit test is good, and regression is returning good results.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1142]: *docummentation = comments

**Comment by aldogonzalez8** [comment]
@ehearty If you make changes before the weekend, can you put the release date on Monday? I prefer not to ship on Thursday afternoon-Friday. If all tests pass and everything looks good, I will approve and later merge that day. Again, just if you do it before Friday. Thanks.

**Comment by ehearty** [comment]
> @ehearty If you make changes before the weekend, can you put the release date on Monday? I prefer not to ship on Thursday afternoon-Friday. If all tests pass and everything looks good, I will approve and later merge that day. Again, just if you do it before Friday. Thanks.

Sorry for the delay. I've specifically blocked off time this week to make those final updates and will have them submitted in the next couple of days.

**Comment by ehearty** [comment]
@aldogonzalez8 Just letting you know that I've jumped back into the PR today and am working on a unit test for the new sort/filter parameters. Hoping to have the changes in by tomorrow, but will set the release date to Monday as requested. Thanks again for your patience.

**Comment by aldogonzalez8** [comment]
> @aldogonzalez8 Just letting you know that I've jumped back into the PR today and am working on a unit test for the new sort/filter parameters. Hoping to have the changes in by tomorrow, but will set the release date to Monday as requested. Thanks again for your patience.

@ehearty That makes sense to me, thanks!

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py
Comment (aldogonzalez8) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py:1144]: Awesome

**Comment by ehearty** [comment]
@aldogonzalez8 hey there - I think someone on the team needs to approve a vercel deployment? I found a bug while creating the unit test and I'm going to run some additional live tests on my own over the weekend, but would love to have this PR go through all the CI checks as well.

**Comment by aldogonzalez8** [comment]
> @aldogonzalez8 hey there - I think someone on the team needs to approve a vercel deployment? I found a bug while creating the unit test and I'm going to run some additional live tests on my own over the weekend, but would love to have this PR go through all the CI checks as well.

Sure, what problem did you find?

**Comment by ehearty** [comment]
> Sure, what problem did you find?

Well, I ran into an issue transitioning my custom connector implementation to the latest airbyte version due to the assumption that "hs_object_id" will always be returned in the search results (which it is for all of my custom connector's use cases, but not the default case for some other entities). To get around this, and to make the solution more flexible, my approach was to use self.primary_key in the sort/filter instead. 

And that worked beautifully until I discovered yesterday that you can't actually sort and filter by "id" when making the API call. It seems that hs doesn't recognize it as a proper field. I've solved it by setting the key to hs_object_id, but only when the primary_key == "id", and still using primary_key to grab last_id from the result set. It works now, but I hate this solution because we're now assuming that "id" will always equal "hs_object_id" - even though I've never seen an instance where that wasn't the case.

What's bugging me is that all the tests passed even when the bug was still there. So I'm concerned that the unit/integration tests aren't properly exercising the incremental search result use case (which might be due to the connector always executing a full refresh when there's no state available). 

Bah. It's weird. If we can get all the CI checks to pass (not sure if you need to re-authorize that?), though, I'll feel a lot better about it. And - as I said before - I'll be running my own tests over the weekend. 

Hope all that made sense. Thanks!

**Comment by ehearty** [comment]
@aldogonzalez8 wanted to let you know that I ran my own tests over the weekend and I think we're good. I'm still seeing a "Vercel" error in the automated checks, though. Is this something that you could approve?

File: docs/integrations/sources/hubspot.md
Comment (aldogonzalez8) [inline docs/integrations/sources/hubspot.md:340]: @ehearty can you fix this? I believe it is part of the Vercel check problem. Well still investigating but in anyway this is wrong, also you can use vim to add your line, it will not mess with the file as other IDEs do.

Comment (aldogonzalez8) [reply docs/integrations/sources/hubspot.md:340]: Fix this = dates are not descending.

Comment (ehearty) [reply docs/integrations/sources/hubspot.md]: @aldogonzalez8 no idea what happened there... but it's fixed now

**Review by aldogonzalez8 — APPROVED** [review]
APPROVED

**Comment by aldogonzalez8** [comment]
> hs_object_id

@ehearty  Okay, I was researching and reviewing some of our expected responses, and I think you are right about the entities' ids being hs_object_id. I am merging this. Thanks for the extra work. I will keep an eye on this.

**Comment by ehearty** [comment]
> > hs_object_id
> 
> @ehearty Okay, I was researching and reviewing some of our expected responses, and I think you are right about the entities' ids being hs_object_id. I am merging this. Thanks for the extra work. I will keep an eye on this.

Thanks so much @aldogonzalez8 ! It's been a pleasure working with you :)

**Pull request merged**
